OpenRA.Mods.Common/Traits/Upgrades/UpgradeOnDamage.cs

```diff
@@ -35,23 +35,35 @@ public class UpgradeOnDamageInfo : ITraitInfo, Requires<UpgradeManagerInfo>
 		public object Create(ActorInitializer init) { return new UpgradeOnDamage(init.Self, this); }
 	}
 
-	public class UpgradeOnDamage : INotifyDamageStateChanged
+	public class UpgradeOnDamage : INotifyDamageStateChanged, INotifyCreated
 	{
 		readonly UpgradeOnDamageInfo info;
 		readonly UpgradeManager um;
+		readonly Health health;
 		bool granted;
 
 		public UpgradeOnDamage(Actor self, UpgradeOnDamageInfo info)
 		{
 			this.info = info;
-			um = self.TraitOrDefault<UpgradeManager>();
+			um = self.Trait<UpgradeManager>();
+			health = self.Trait<Health>();
 		}
 
-		void INotifyDamageStateChanged.DamageStateChanged(Actor self, AttackInfo e)
+		void INotifyCreated.Created(Actor self)
 		{
-			if (um == null)
-				return;
+			if (!granted && info.ValidDamageStates.HasFlag(health.DamageState))
```

Comment:
To be honest, I don't follow, at least not completely. Do you just mean to put the granting stuff into its own method and then just call that method here and in `INotifyDamageStateChanged.DamageStateChanged`, or should `Created` trigger a fake `DamageStateChanged` (pretending it changed)?